sequencer/src/block.rs

```diff
@@ -81,45 +69,147 @@ pub struct Metadata {
-            .u64_field("timestamp", self.timestamp())
-            .u64_field("l1_head", self.l1_head())
-            .array_field(
-                "l1_finalized",
-                self.l1_finalized()
-                    .iter()
-                    .map(|block| block.commit())
-                    .collect::<Vec<_>>()
-                    .as_slice(),
-            )
+impl Committable for Header {
+    fn commit(&self) -> Commitment<Self> {
+        RawCommitmentBuilder::new(&Self::tag())
+            .u64_field("height", self.height)
+            .u64_field("timestamp", self.timestamp)
+            .u64_field("l1_head", self.l1_head)
+            .optional("l1_finalized", &self.l1_finalized)
+            .constant_str("payload_commitment")
+            .fixed_size_bytes(&self.payload_commitment.into())
             .field("transactions_root", self.transactions_root.commit())
             .finalize()
     }
+
+    fn tag() -> String {
+        "BLOCK".into()
+    }
 }
 
-#[derive(Clone, Debug, Deserialize, Serialize, Hash, PartialEq, Eq)]
-pub struct Block {
-    metadata: Metadata,
+impl Committable for NMTRoot {
+    fn commit(&self) -> Commitment<Self> {
+        let mut comm_bytes = vec![];
+        self.root
+            .serialize_with_mode(&mut comm_bytes, ark_serialize::Compress::Yes)
+            .unwrap();
+        RawCommitmentBuilder::new(&Self::tag())
+            .var_size_field("root", &comm_bytes)
+            .finalize()
+    }
 
-    #[serde(with = "nmt_serializer")]
-    pub(crate) transaction_nmt: TransactionNMT,
+    fn tag() -> String {
+        "NMTROOT".into()
+    }
 }
 
-impl Block {
-    pub fn header(&self) -> Header {
-        Header {
-            metadata: self.metadata.clone(),
-            transactions_root: NMTRoot {
-                root: self.transaction_nmt.commitment().digest(),
-            },
+impl BlockHeader for Header {
+    type Payload = Payload;
+
+    fn new(payload_commitment: VidCommitment, metadata: NMTRoot, parent_header: &Self) -> Self {
+        Self::from_parent(
+            payload_commitment,
+            metadata,
+            Some(parent_header.height),
+            parent_header.timestamp,
+        )
+    }
+
+    fn genesis() -> (
+        Self,
+        Self::Payload,
+        <Self::Payload as BlockPayload>::Metadata,
+    ) {
+        let (payload, metadata) = Payload::genesis();
+        let payload_commitment = vid_commitment(&payload.encode().unwrap().collect());
+        let header = Self::from_parent(payload_commitment, metadata, None, 0);
+        (header, payload, metadata)
+    }
+
+    fn block_number(&self) -> u64 {
+        self.height
+    }
+
+    fn payload_commitment(&self) -> VidCommitment {
+        self.payload_commitment
+    }
+
+    fn metadata(&self) -> NMTRoot {
+        self.transactions_root
+    }
+}
+
+impl Header {
+    fn from_parent(
+        payload_commitment: VidCommitment,
+        transactions_root: NMTRoot,
+        parent_height: Option<u64>,
```

Comment:
Nit: sucks that we need `Option` just to handle the genesis block. Couldn't we just construct the genesis block directly without calling `from_parent`, thus avoiding the need for `Option` here?

Reply:
Sure. I wanted to avoid duplicating the logic for getting the L1 snapshot, since it's a bit complex. But all the rest of the complexity (computing the height, adjusting the timestamp) doesn't apply to the genesis block. So there is probably a better way to structure this.

Reply:
Realized this is a bug anyways. Everyone needs to have the same genesis header, so this needs to be deterministic. I think it is only down to luck and the 1-second timestamp resolution that the tests pass

Reply:
Presumably we'll eventually hard-code the genesis block or read it from disk, no? I guess that could get tricky if we have wall-clock sync constraints.